doc/luaotfload.conf.5
.\" Man page generated from reStructuredText.
.
.TH LUAOTFLOAD.CONF 5 "2019-10-29" "3.0008-dev" "text processing"
.SH NAME
luaotfload.conf \- Luaotfload configuration file
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.INDENT 0.0
.IP \(bu 2
\fB\&./luaotfload{.conf,rc}\fP
.IP \(bu 2
\fBXDG_CONFIG_HOME/luaotfload/luaotfload{.conf,rc}\fP
.IP \(bu 2
\fB~/.luaotfloadrc\fP
.UNINDENT
.SH DESCRIPTION
.sp
The file \fBluaotfload.conf\fP contains configuration options for
\fILuaotfload\fP, a font loading and font management component for LuaTeX.
.SH EXAMPLE
.sp
A small Luaotfload configuration file with few customizations could
look as follows:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
[db]
    formats = afm,ttf
    compress = false

[misc]
    termwidth = 60

[run]
    log\-level = 6
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
This will make Luaotfload ignore all font files except for PostScript
binary fonts with a matching AFM file, and Truetype fonts. Also, an
uncompressed index file will be dumped which is going to be much larger
than the default gzip’ed index. The terminal width
is truncated to 60 characters which influences the verbose output
during indexing. Finally, the verbosity is increased greatly: each font
file being processed will be printed to the stdout on a separate line,
along with lots of other information.
.sp
To observe the difference in behavior, save above snippet to
\fB\&./luaotfload.conf\fP and update the font index:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
luaotfload\-tool \-\-update \-\-force
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The current configuration can be written to disk using
\fBluaotfload\-tool\fP:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
luaotfload\-tool \-\-dumpconf > luaotfload.conf
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The result can itself be used as a configuration file.
.SH SYNTAX
.sp
The configuration file syntax follows the common INI format. For a more
detailed description please refer to the section “CONFIGURATION FILE”
of \fBgit\-config\fP(1). A brief list of rules is given below:
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.IP \(bu 2
Blank lines and lines starting with a semicolon (\fB;\fP) are ignored.
.IP \(bu 2
A configuration file is partitioned into sections that are declared
by specifying the section title in brackets on a separate line:
.INDENT 2.0
.INDENT 3.5
.sp
.nf
.ft C
[some\-section]
\&... section content ...
.ft P
.fi
.UNINDENT
.UNINDENT
.IP \(bu 2
Sections consist of one or more variable assignments of the form
\fBvariable\-name = value\fP  E. g.:
.INDENT 2.0
.INDENT 3.5
.sp
.nf
.ft C
[foo]
    bar = baz
    quux = xyzzy
    ...
.ft P
.fi
.UNINDENT
.UNINDENT
.IP \(bu 2
Section and variable names may contain only uppercase and lowercase
letters as well as dashes (\fB\-\fP).
.UNINDENT
.UNINDENT
.UNINDENT
.SH VARIABLES
.sp
Variables in belong into a configuration section and their values must
be of a certain type. Some of them have further constraints. For
example, the “color callback” must be a string of one of the values
\fBpost_linebreak_filter\fP, \fBpre_linebreak_filter\fP, or
\fBpre_output_filter\fP, defined in the section \fIrun\fP of the
configuration file.
.sp
Currently, the configuration is organized into four sections:
.INDENT 0.0
.TP
.B db
Options relating to the font index.
.TP
.B misc
Options without a clearly defined category.
.TP
.B paths
Path and file name settings.
.TP
.B run
Options controlling runtime behavior of Luaotfload.
.UNINDENT
.sp
The list of valid variables, the sections they are part of and their
type is given below. Types represent Lua types that the values must be
convertible to; they are abbreviated as follows: \fBs\fP for the \fIstring\fP
type, \fBn\fP for \fInumber\fP, \fBb\fP for \fIboolean\fP\&. A value of \fBnil\fP means
the variable is unset.
.SS Section \fBdb\fP
.TS
center;
|l|l|l|.
_
T{
variable
T}	T{
type
T}	T{
default
T}
_
T{
compress
T}	T{
b
T}	T{
\fBtrue\fP
T}
_
T{
designsize\-dimen
T}	T{
b
T}	T{
\fBbp\fP
T}
_
T{
formats
T}	T{
s
T}	T{
\fB"otf,ttf,ttc"\fP
T}
_
T{
max\-fonts
T}	T{
n
T}	T{
\fB2^51\fP
T}
_
T{
scan\-local
T}	T{
b
T}	T{
\fBfalse\fP
T}
_
T{
skip\-read
T}	T{
b
T}	T{
\fBfalse\fP
T}
_
T{
strip
T}	T{
b
T}	T{
\fBtrue\fP
T}
_
T{
update\-live
T}	T{
b
T}	T{
\fBtrue\fP
T}
_
.TE
.sp
The flag \fBcompress\fP determines whether the font index (usually
\fBluaotfload\-names.lua[.gz]\fP will be stored in compressed forms.
If unset it is equivalent of passing \fB\-\-no\-compress\fP to
\fBluaotfload\-tool\fP\&. Since the file is only created for convenience
and has no effect on the runtime behavior of Luaotfload, the flag
should remain set. Most editors come with zlib support anyways.
.sp
The setting \fBdesignsize\-dimen\fP applies when looking up fonts from
families with design sizes. In Opentype, these are specified as
“decipoints” where one decipoint equals ten DTP style “big points”.
When indexing fonts these values are converted to \fBsp\fP\&. In order to
treat the values as though they were specified in TeX points or Didot
points, set \fBdesignsize\-dimen\fP to \fBpt\fP or \fBdd\fP\&.
.sp
The list of \fBformats\fP must be a comma separated sequence of strings
containing one or more of these elements:
.INDENT 0.0
.IP \(bu 2
\fBotf\fP               (OpenType format),
.IP \(bu 2
\fBttf\fP and \fBttc\fP       (TrueType format),
.IP \(bu 2
\fBafm\fP               (Adobe Font Metrics),
.UNINDENT
.sp
It corresponds loosely to the \fB\-\-formats\fP option to
\fBluaotfload\-tool\fP\&. Invalid or duplicate members are ignored; if the
list does not contain any useful identifiers, the default list
\fB"otf,ttf,ttc"\fP will be used.
.sp
The variable \fBmax\-fonts\fP determines after processing how many font
files the font scanner will terminate the search. This is useful for
debugging issues with the font index and has the same effect as the
option \fB\-\-max\-fonts\fP to \fBluaotfload\-tools\fP\&.
.sp
The \fBscan\-local\fP flag, if set, will incorporate the current working
directory as a font search location. NB: This will potentially slow
down document processing because a font index with local fonts will not
be saved to disk, so these fonts will have to be re\-indexed whenever
the document is built.
.sp
The \fBskip\-read\fP flag is only useful for debugging: It makes
Luaotfload skip reading fonts. The font information for rebuilding the
index is taken from the presently existing one.
.sp
Unsetting the \fBstrip\fP flag prevents Luaotfload from removing data
from the index that is only useful when processing font files. NB: this
can increase the size of the index files significantly and has no
effect on the runtime behavior.
.sp
If \fBupdate\-live\fP is set, Luaotfload will reload the database if it
cannot find a requested font. Those who prefer to update manually using
\fBluaotfload\-tool\fP should unset this flag. This option does not affect
rebuilds due to version mismatch.
.SS Section \fBdefault\-features\fP
.sp
By default Luaotfload enables \fBnode\fP mode and picks the default font
features that are prescribed in the OpenType standard. This behavior
may be overridden in the \fBdefault\-features\fP section. Global defaults
that will be applied for all scripts can be set via the \fBglobal\fP
option, others by the script they are to be applied to. For example,
a setting of
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
[default\-features]
    global = mode=base,color=0000FF
    dflt   = smcp,onum
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
would force \fIbase\fP mode, tint all fonts blue and activate small
capitals and text figures globally. Features are specified as a comma

NEWS
Change History
--------------
2019-11-04 luaotfload v3.1
    * added code to use the harfbuzz library of luahbtex ("mode=harf")
    * fonts in ttc can now be indexed by name
    * remove a number of lua tables from the global enviroment (_G) to reduce the polution. 
      Kept there are only the tables luaotfload, fonts and nodes.
    * sync with context files from 2019-10-29

2019-09-13 luaotfload v3.0
    * invisible option to avoid that some glyph are shown (issue #63)
    * correct small bug in xheight calculation 

2019-08-11 luaotfload v2.99
    * some clean up
    * fix a bug in luaotfload.aux.resolve_fontname()  (issue #76)
    * removed dependency on slnunicode (PR#72) 
    * sync with context files from 2019-08-11

2019-06-20 luaotfload v2.98
    * add embolden key
    * add missing char detection issue #65
    * sync with context files from 2019-07-04
    
2019-05-18 luaotfload v2.97
    * fix issue #47
    * fix whatsits interfering with letterspacing (issue #53)
    * fix luaotfload-tool switches version and find not working correctly (PR#59)
    * fix luaotfload-tool support of ttc fonts (PR#58)
    * sync with context files from 2019-05-18 (improves handling of large fonts, see e.g. issue #55 and PR#58)
    
2019-01-28 luaotfload v2.96
    * repaired broken letterspace (issue #38)
    * changed handling of spaces in letterspacing
    * changed handling of ligatures in letterspacing
    * corrected detection of bold fonts (issue #41)

2019-01-28 luaotfload v2.95
    * imported context files of 2019-01-28
    * changed bold indification (for lucida demi bold) (issue #33)
    * changed handling of tex file (issue #35)
    * changed resolver code (pr #36)

2019-01-14 luaotfload v2.94
    * imported context files of 2018-12-19
    * removed luaotfload-characters from init - use the basics-chr instead.
    * allow custom fontloaders through the mode-key https://github.com/u-fischer/luaotfload/pull/25
    * resolved issue #29

2018-10-28 luaotfload v2.93
    * reinserted forgotten man files
    * unified and improved version/module info for automatic tagging


2018-10-21 luaotfload v2.92
    * better deva support, issue #9
    * test for --safer and abort, issue #12 
    * newest context files   
    * correct file: syntax for tfm files, issue #11.    

2018-10-03 luaotfload v2.91
    * solved copy & paste of variant glyphs, https://tex.stackexchange.com/questions/451442/localized-glyphs-locl-have-unicode-value-ufffd
    * solved problem with PUA no longer accessible with \char, see https://github.com/u-fischer/luaotfload/issues/7

2018-09-21 luaotfload v2.9 
    * added various testfiles for the l3build system/travis tests to the github repository
    * solved a bug regarding directions https://github.com/u-fischer/lua-font-pond/issues/12
    * adapted luaotfload-letterspace to new fontloader
    * solved the case problem see https://github.com/u-fischer/lua-font-pond/issues/6
    * Fix parsing of font.conf see https://github.com/lualatex/luaotfload/pull/420
    * imported new versions of the context files - fontloader is compatible with luatex 1.08 again
    * adapted the files list 
    * created a luaotfload-filelist.lua which describes the files of the package
    * removed some not functional files 
    * adapted mkstatus to use the new file list.
    
2017-01-29, luaotfload v2.8:
    * Latest fontloader code.
    * Support for Luatex 1.0.
    * Integration of the extended AFM handler.
    * Fixes to font family assignment.
    * Skip initialization if Luaotfload is already loaded.
    * Optical sizes advertised by fonts treated as big points.

2016/04/21, luaotfload v2.7:
    * Rework lookup chaining.
    * Combining glyphs from different fonts (``combo: ...`` requests).
    * Glyph fallbacks.
    * Latest fontloader code.
    * Adapt packaging to changed upstream file layout.
    * Remove support for builtin Fontforge libraries (this includes the PFA,
      PFB, DFONT, and feature file readers).
    * Allow configuration of anonymous lookups.

2015/12/09, luaotfload v2.6

luaotfload.lua
-----------------------------------------------------------------------
--         FILE:  luaotfload.lua
--  DESCRIPTION:  Luatex fontloader initialization
-- REQUIREMENTS:  luatex v.0.78 or later, the lualibs package
--       AUTHOR:  Élie Roux, Khaled Hosny, Philipp Gesang
--      VERSION:  same as Luaotfload
--     MODIFIED:  2014-01-16 06:51:20+0100
-----------------------------------------------------------------------
--
--- Note:
--- This file was part of the original luaotfload.dtx and has been
--- converted to a pure Lua file during the transition from Luaotfload
--- version 2.4 to 2.5. Thus, the comments are still in TeX (Latex)
--- markup.

if not modules then modules = { } end modules ["luaotfload"] = {
    version   = "2.5",
    comment   = "fontloader initialization",
    author    = "Hans Hagen, Khaled Hosny, Elie Roux, Philipp Gesang",
    copyright = "PRAGMA ADE / ConTeXt Development Team",
    license   = "GNU General Public License v. 2.0"
}


--[[doc--

    This file initializes the system and loads the font loader. To
    minimize potential conflicts between other packages and the code
    imported from \CONTEXT, several precautions are in order. Some of
    the functionality that the font loader expects to be present, like
    raw access to callbacks, are assumed to have been disabled by
    \identifier{luatexbase} when this file is processed. In some cases
    it is possible to trick it by putting dummies into place and
    restoring the behavior from \identifier{luatexbase} after
    initilization. Other cases such as attribute allocation require
    that we hook the functionality from \identifier{luatexbase} into
    locations where they normally wouldn’t be.

    Anyways we can import the code base without modifications, which is
    due mostly to the extra effort by Hans Hagen to make \LUATEX-Fonts
    self-contained and encapsulate it, and especially due to his
    willingness to incorporate our suggestions.

--doc]]--

luaotfload                        = luaotfload or { }
local luaotfload                  = luaotfload

config                            = config or { }
config.luaotfload                 = config.luaotfload or { }
------.luaotfload.resolver        = config.luaotfload.resolver         or "normal"
config.luaotfload.resolver        = config.luaotfload.resolver         or "cached"
config.luaotfload.definer         = config.luaotfload.definer          or "patch"
config.luaotfload.loglevel        = config.luaotfload.loglevel         or 2
config.luaotfload.color_callback  = config.luaotfload.color_callback   or "pre_linebreak_filter"
config.luaotfload.prioritize      = config.luaotfload.prioritize       or "sys"
config.luaotfload.names_dir       = config.luaotfload.names_dir        or "names"
config.luaotfload.cache_dir       = config.luaotfload.cache_dir        or "fonts"
config.luaotfload.index_file      = config.luaotfload.index_file       or "luaotfload-names.lua"
config.luaotfload.formats         = config.luaotfload.formats          or "otf,ttf,ttc,dfont"

if not config.luaotfload.strip then
    config.luaotfload.strip = true
end

luaotfload.module = {
    name          = "luaotfload",
    version       = 2.50000,
    date          = "2014/**/**",
    description   = "OpenType layout system.",
    author        = "Elie Roux & Hans Hagen",
    copyright     = "Elie Roux",
    license       = "GPL v2.0"
}

local luatexbase       = luatexbase

local setmetatable     = setmetatable
local type, next       = type, next

local kpsefind_file    = kpse.find_file
local lfsisfile        = lfs.isfile

local add_to_callback, create_callback =
      luatexbase.add_to_callback, luatexbase.create_callback
local reset_callback, call_callback =
      luatexbase.reset_callback, luatexbase.call_callback

local dummy_function = function () end

local error, warning, info, log =
    luatexbase.provides_module(luaotfload.module)

luaotfload.error        = error
luaotfload.warning      = warning
luaotfload.info         = info
luaotfload.log          = log

--[[doc--

     We set the minimum version requirement for \LUATEX to v0.76,
     because the font loader requires recent features like direct
     attribute indexing and \luafunction{node.end_of_math()} that aren’t
     available in earlier versions.\footnote{%
      See Taco’s announcement of v0.76:
      \url{http://comments.gmane.org/gmane.comp.tex.luatex.user/4042}
      and this commit by Hans that introduced those features.
      \url{http://repo.or.cz/w/context.git/commitdiff/a51f6cf6ee087046a2ae5927ed4edff0a1acec1b}.
    }

--doc]]--

local luatex_version = 76

if tex.luatexversion < luatex_version then
    warning("LuaTeX v%.2f is old, v%.2f is recommended.",
             tex.luatexversion/100,
             luatex_version   /100)
    --- we install a fallback for older versions as a safety
    if not node.end_of_math then
        local math_t          = node.id"math"
        local traverse_nodes  = node.traverse_id
        node.end_of_math = function (n)
            for n in traverse_nodes(math_t, n.next) do
                return n
            end
        end
    end
end

--[[doc--

    \subsection{Module loading}
    We load the files imported from \CONTEXT with this function. It
    automatically prepends the prefix \fileent{luaotfload-} to its
    argument, so we can refer to the files with their actual \CONTEXT
    name.

--doc]]--

local fl_prefix = "luaotfload" -- “luatex” for luatex-plain
local loadmodule = function (name)
    require(fl_prefix .."-"..name)
end

--[[doc--

  Before \TeX Live 2013 version, \LUATEX had a bug that made ofm fonts
  fail when called with their extension. There was a side-effect making
  ofm totally unloadable when luaotfload was present. The following
  lines are a patch for this bug. The utility of these lines is
  questionable as they are not necessary since \TeX Live 2013. They
  should be removed in the next version.

--doc]]--


local p_dot, p_slash = P".",  P"/"
local p_suffix       = (p_dot * (1 - p_dot - p_slash)^1 * P(-1)) / ""
local p_removesuffix = Cs((p_suffix + 1)^1)

local find_vf_file = function (name)
    local fullname = kpsefind_file(name, "ovf")
    if not fullname then
        --fullname = kpsefind_file(file.removesuffix(name), "ovf")
        fullname = kpsefind_file(lpegmatch(p_removesuffix, name), "ovf")
    end
    if fullname then
        log("loading virtual font file %s.", fullname)
    end
    return fullname
end

--[[doc--

    \subsection{Preparing the Font Loader}
    We treat the fontloader as a black box so behavior is consistent
    between formats.
    We load the fontloader code directly in the same fashion as the
    Plain format \identifier{luatex-fonts} that is part of Context.
    How this is executed depends on the presence on the
    \emphasis{merged font loader code}.
    In \identifier{luaotfload} this is contained in the file
    \fileent{luaotfload-merged.lua}.
    If this file cannot be found, the original libraries from \CONTEXT
    of which the merged code was composed are loaded instead.
    Since these files are not shipped with Luaotfload, an installation
    of Context is required.
    (Since we pull the fontloader directly from the Context minimals,
    the necessary Context version is likely to be more recent than that
    of other TeX distributions like Texlive.)
    The imported font loader will call \luafunction{callback.register}
    once while reading \fileent{font-def.lua}.
    This is unavoidable unless we modify the imported files, but
    harmless if we make it call a dummy instead.
    However, this problem might vanish if we decide to do the merging
    ourselves, like the \identifier{lualibs} package does.
    With this step we would obtain the freedom to load our own
    overrides in the process right where they are needed, at the cost
    of losing encapsulation.
    The decision on how to progress is currently on indefinite hold.

--doc]]--

local starttime = os.gettimeofday()

local trapped_register  = callback.register
callback.register       = dummy_function

--[[doc--

    By default, the fontloader requires a number of \emphasis{private
    attributes} for internal use.
    These must be kept consistent with the attribute handling methods
    as provided by \identifier{luatexbase}.
    Our strategy is to override the function that allocates new
    attributes before we initialize the font loader, making it a
    wrapper around \luafunction{luatexbase.new_attribute}.\footnote{%
        Many thanks, again, to Hans Hagen for making this part
        configurable!
    }
    The attribute identifiers are prefixed “\fileent{luaotfload@}” to
    avoid name clashes.

--doc]]--

do
    local new_attribute    = luatexbase.new_attribute
    local the_attributes   = luatexbase.attributes

    attributes = attributes or { }

    attributes.private = function (name)
        local attr   = "luaotfload@" .. name --- used to be: “otfl@”
        local number = the_attributes[attr]
        if not number then
            number = new_attribute(attr)
        end
        return number
    end
end

--[[doc--

    These next lines replicate the behavior of
    \fileent{luatex-fonts.lua}.

--doc]]--

local context_environment = { }

local push_namespaces = function ()
    log("push namespace for font loader")
    local normalglobal = { }
    for k, v in next, _G do
        normalglobal[k] = v
    end
    return normalglobal
end

local pop_namespaces = function (normalglobal, isolate)
    if normalglobal then
        local _G = _G
        local mode = "non-destructive"
        if isolate then mode = "destructive" end
        log("pop namespace from font loader -- " .. mode)
        for k, v in next, _G do
            if not normalglobal[k] then
                context_environment[k] = v
                if isolate then
                    _G[k] = nil
                end
            end
        end
        for k, v in next, normalglobal do
            _G[k] = v
        end
        -- just to be sure:
        setmetatable(context_environment,_G)
    else
        log("irrecoverable error during pop_namespace: no globals to restore")
        os.exit()
    end
end

luaotfload.context_environment  = context_environment
luaotfload.push_namespaces      = push_namespaces
luaotfload.pop_namespaces       = pop_namespaces

local our_environment = push_namespaces()

--[[doc--

    The font loader requires that the attribute with index zero be
    zero. We happily oblige.
    (Cf. \fileent{luatex-fonts-nod.lua}.)

--doc]]--

tex.attribute[0] = 0

--[[doc--

    Now that things are sorted out we can finally load the fontloader.

--doc]]--

loadmodule "fontloader.lua"
---loadmodule"font-odv.lua" --- <= Devanagari support from Context

if fonts then

    if not fonts._merge_loaded_message_done_ then
        log [["I am using the merged version of 'luaotfload.lua' here.]]
        log [[ If you run into problems or experience unexpected]]
        log [[ behaviour, and if you have ConTeXt installed you can try]]
        log [[ to delete the file 'luaotfload-merged.lua' as I might]]
        log [[ then use the possibly updated libraries. The merged]]
        log [[ version is not supported as it is a frozen instance.]]
        log [[ Problems can be reported to the ConTeXt mailing list."]]
    end
    fonts._merge_loaded_message_done_ = true

else--- the loading sequence is known to change, so this might have to
    --- be updated with future updates!
    --- do not modify it though unless there is a change to the merged
    --- package!
    loadmodule("l-lua.lua")
    loadmodule("l-lpeg.lua")
    loadmodule("l-function.lua")
    loadmodule("l-string.lua")
    loadmodule("l-table.lua")
    loadmodule("l-io.lua")
    loadmodule("l-file.lua")
    loadmodule("l-boolean.lua")
    loadmodule("l-math.lua")
    loadmodule("util-str.lua")
    loadmodule('luatex-basics-gen.lua')
    loadmodule('data-con.lua')
    loadmodule('luatex-basics-nod.lua')
    loadmodule('font-ini.lua')
    loadmodule('font-con.lua')
    loadmodule('luatex-fonts-enc.lua')
    loadmodule('font-cid.lua')
    loadmodule('font-map.lua')
    loadmodule('luatex-fonts-syn.lua')
    loadmodule('luatex-fonts-tfm.lua')
    loadmodule('font-oti.lua')
    loadmodule('font-otf.lua')
    loadmodule('font-otb.lua')
    loadmodule('luatex-fonts-inj.lua')  --> since 2014-01-07, replaces node-inj.lua
    loadmodule('font-ota.lua')
    loadmodule('luatex-fonts-otn.lua')  --> since 2014-01-07, replaces font-otn.lua
    loadmodule('font-otp.lua')          --> since 2013-04-23
    loadmodule('luatex-fonts-lua.lua')
    loadmodule('font-def.lua')
    loadmodule('luatex-fonts-def.lua')
    loadmodule('luatex-fonts-ext.lua')
    loadmodule('luatex-fonts-cbk.lua')
end --- non-merge fallback scope

--[[doc--

    Here we adjust the globals created during font loader
    initialization. If the second argument to
    \luafunction{pop_namespaces()} is \verb|true| this will restore the
    state of \luafunction{_G}, eliminating every global generated since
    the last call to \luafunction{push_namespaces()}. At the moment we
    see no reason to do this, and since the font loader is considered
    an essential part of \identifier{luatex} as well as a very well
    organized piece of code, we happily concede it the right to add to
    \luafunction{_G} if needed.

--doc]]--

pop_namespaces(our_environment, false)-- true)

log("fontloader loaded in %0.3f seconds", os.gettimeofday()-starttime)

--[[doc--

    \subsection{Callbacks}
    After the fontloader is ready we can restore the callback trap from
    \identifier{luatexbase}.

--doc]]--

callback.register = trapped_register

--[[doc--

    We do our own callback handling with the means provided by
    luatexbase.
    Note: \luafunction{pre_linebreak_filter} and
    \luafunction{hpack_filter} are coupled in \CONTEXT in the concept
    of \emphasis{node processor}.

--doc]]--

add_to_callback("pre_linebreak_filter",
                nodes.simple_font_handler,
                "luaotfload.node_processor",
                1)
add_to_callback("hpack_filter",
                nodes.simple_font_handler,
                "luaotfload.node_processor",
                1)
add_to_callback("find_vf_file",
                find_vf_file, "luaotfload.find_vf_file")

loadmodule "override.lua"   --- “luat-ovr”

logs.set_loglevel(config.luaotfload.loglevel)

--[[doc--

    Now we load the modules written for \identifier{luaotfload}.

--doc]]--
loadmodule "parsers.lua"    --- new in 2.5; fonts.conf and syntax
loadmodule "loaders.lua"    --- “font-pfb” new in 2.0, added 2011
loadmodule "database.lua"   --- “font-nms”
loadmodule "colors.lua"     --- “font-clr”

--[[doc--

    Relying on the \verb|name:| resolver for everything has been the
    source of permanent trouble with the database.
    With the introduction of the new syntax parser we now have enough
    granularity to distinguish between the \XETEX emulation layer and
    the genuine \verb|name:| and \verb|file:| lookups of \LUATEX-Fonts.
    Another benefit is that we can now easily plug in or replace new
    lookup behaviors if necessary.
    The name resolver remains untouched, but it calls
    \luafunction{fonts.names.resolve()} internally anyways (see
    \fileent{luaotfload-database.lua}).

--doc]]--

local filesuffix          = file.suffix
local fileremovesuffix    = file.removesuffix
local request_resolvers   = fonts.definers.resolvers
local formats             = fonts.formats
local names               = fonts.names
formats.ofm               = "type1"

fonts.encodings.known     = fonts.encodings.known or { }

--[[doc--

    \identifier{luaotfload} promises easy access to system fonts.
    Without additional precautions, this cannot be achieved by
    \identifier{kpathsea} alone, because it searches only the
    \fileent{texmf} directories by default.
    Although it is possible for \identifier{kpathsea} to include extra
    paths by adding them to the \verb|OSFONTDIR| environment variable,
    this is still short of the goal »\emphasis{it just works!}«.
    When building the font database \identifier{luaotfload} scans
    system font directories anyways, so we already have all the
    information for looking sytem fonts.
    With the release version 2.2 the file names are indexed in the
    database as well and we are ready to resolve \verb|file:| lookups
    this way.
    Thus we no longer need to call the \identifier{kpathsea} library in
    most cases when looking up font files, only when generating the
    database, and when verifying the existence of a file in the
    \fileent{texmf} tree.

--doc]]--

local resolve_file        = names.crude_file_lookup
--local resolve_file        = names.crude_file_lookup_verbose
local resolve_name        = names.resolve_name

local file_resolver = function (specification)
    local name    = resolve_file (specification.name)
    local suffix  = filesuffix(name)
    if formats[suffix] then
        specification.forced      = suffix
        specification.forcedname  = file.removesuffix(name)
    else
        specification.name = name
    end
end

request_resolvers.file = file_resolver

--[[doc--

    We classify as \verb|anon:| those requests that have neither a
    prefix nor brackets. According to Khaled\footnote{%
        \url{https://github.com/phi-gamma/luaotfload/issues/4#issuecomment-17090553}.
    }
    they are the \XETEX equivalent of a \verb|name:| request, so we
    will be treating them as such.

--doc]]--

--request_resolvers.anon = request_resolvers.name

--[[doc--

    There is one drawback, though.
    This syntax is also used for requesting fonts in \identifier{Type1}
    (\abbrev{tfm}, \abbrev{ofm}) format.
    These are essentially \verb|file:| lookups and must be caught
    before the \verb|name:| resolver kicks in, lest they cause the
    database to update.
    Even if we were to require the \verb|file:| prefix for all
    \identifier{Type1} requests, tests have shown that certain fonts
    still include further fonts (e.~g. \fileent{omlgcb.ofm} will ask
    for \fileent{omsecob.tfm}) \emphasis{using the old syntax}.
    For this reason, we introduce an extra check with an early return.

--doc]]--

local type1_formats = { "tfm", "ofm", }

request_resolvers.anon = function (specification)
    local name = specification.name
    for i=1, #type1_formats do
        local format = type1_formats[i]
        if resolvers.findfile(name, format) then
            specification.forcedname = file.addsuffix(name, format)
            specification.forced     = format
            return
        end
    end
    --- under some weird circumstances absolute paths get
    --- passed to the definer; we have to catch them
    --- before the name: resolver misinterprets them.
    name = specification.specification
    local exists, _ = lfsisfile(name)
    if exists then --- garbage; we do this because we are nice,
                   --- not because it is correct
        logs.names_report("log", 1, "load", "file %q exists", name)
        logs.names_report("log", 1, "load",
          "... overriding borked anon: lookup with path: lookup")
        specification.name = name
        request_resolvers.path(specification)
        return
    end
    request_resolvers.name(specification)
end

--[[doc--

    Prior to version 2.2, \identifier{luaotfload} did not distinguish
    \verb|file:| and \verb|path:| lookups, causing complications with
    the resolver.
    Now we test if the requested name is an absolute path in the file
    system, otherwise we fall back to the \verb|file:| lookup.

--doc]]--

request_resolvers.path = function (specification)
    local name       = specification.name
    local exists, _  = lfsisfile(name)
    if not exists then -- resort to file: lookup
        logs.names_report("log", 1, "load",
          "path lookup of %q unsuccessful, falling back to file:",
          name)
        file_resolver (specification)
    else
        local suffix = filesuffix (name)
        if formats[suffix] then
            specification.forced      = suffix
            specification.name        = file.removesuffix(name)
            specification.forcedname  = name
        else
            specification.name = name
        end
    end
end

--[[doc--

    {\bfseries EXPERIMENTAL}:
    \identifier{kpse}-only resolver, for those who can do without
    system fonts.

--doc]]--

request_resolvers.kpse = function (specification)
    local name       = specification.name
    local suffix     = filesuffix(name)
    if suffix and formats[suffix] then
        name = file.removesuffix(name)
        if resolvers.findfile(name, suffix) then
            specification.forced       = suffix
            specification.forcedname   = name
            return
        end
    end
    for t, format in next, formats do --- brute force
        if kpse.find_file (name, format) then
            specification.forced = t
            specification.name   = name
            return
        end
    end
end

--[[doc--

    The \verb|name:| resolver wraps the database function
    \luafunction{resolve_name}.

--doc]]--

--- fonts.names.resolvers.name -- Customized version of the
--- generic name resolver.

request_resolvers.name = function (specification)
    local resolved, subfont = resolve_name (specification)
    if resolved then
        specification.resolved   = resolved
        specification.sub        = subfont
        specification.forced     = filesuffix (resolved)
        specification.forcedname = resolved
        specification.name       = fileremovesuffix (resolved)
    else
        file_resolver (specification)
    end
end

--[[doc--

    Also {\bfseries EXPERIMENTAL}: custom file resolvers via callback.

--doc]]--
create_callback("luaotfload.resolve_font", "simple", dummy_function)

request_resolvers.my = function (specification)
    call_callback("luaotfload.resolve_font", specification)
end

--[[doc--

    We create a callback for patching fonts on the fly, to be used by
    other packages.
    It initially contains the empty function that we are going to
    override below.

--doc]]--

create_callback("luaotfload.patch_font", "simple", dummy_function)

--[[doc--

    \subsection{\CONTEXT override}
    \label{define-font}
    We provide a simplified version of the original font definition
    callback.

--doc]]--

local read_font_file = fonts.definers.read

--- spec -> size -> id -> tmfdata
local patch_defined_font = function (specification, size, id)
    local tfmdata = read_font_file(specification, size, id)
    if type(tfmdata) == "table" and tfmdata.shared then
        --- We need to test for the “shared” field here
        --- or else the fontspec capheight callback will
        --- operate on tfm fonts.
        call_callback("luaotfload.patch_font", tfmdata, specification)
    end
    return tfmdata
end

reset_callback "define_font"

--[[doc--

    Finally we register the callbacks.

--doc]]--

local font_definer = config.luaotfload.definer

if font_definer == "generic"  then
  add_to_callback("define_font",
                  fonts.definers.read,
                  "luaotfload.define_font",
                  1)
elseif font_definer == "patch"  then
  add_to_callback("define_font",
                  patch_defined_font,
                  "luaotfload.define_font",
                  1)
end

loadmodule "features.lua"     --- contains what was “font-ltx” and “font-otc”
loadmodule "letterspace.lua"  --- extra character kerning
loadmodule "auxiliary.lua"    --- additionaly high-level functionality (new)

luaotfload.aux.start_rewrite_fontname () --- to be migrated to fontspec

-- vim:tw=79:sw=4:ts=4:et
